### PR TITLE
BUG: alignment of CategoricalIndex

### LINKED ISSUE #13365: Alignment of CategoricalIndex
#### Code Sample, a copy-pastable example if possible

```
In [2]: c = pd.CategoricalIndex(['3z53', '3z53', 'LoJG', 'LoJG', 'LoJG', 'N503'])

In [3]: s1 = pd.Series(2, index=c)

In [4]: s2 = pd.Series(2, index=c[:-1])

In [5]: s1*s2
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-7426096cfe41> in <module>()
----> 1 s1*s2

/home/pietro/nobackup/repo/pandas/pandas/core/ops.py in wrapper(left, right, name, na_op)
    647             else:
    648                 index, lidx, ridx = left.index.join(rindex, how='outer',
--> 649                                                     return_indexers=True)
    650 
    651                 if lidx is not None:

/home/pietro/nobackup/repo/pandas/pandas/indexes/base.py in join(self, other, how, level, return_indexers)
   2524         if not self.is_unique and not other.is_unique:
   2525             return self._join_non_unique(other, how=how,
-> 2526                                          return_indexers=return_indexers)
   2527         elif not self.is_unique or not other.is_unique:
   2528             if self.is_monotonic and other.is_monotonic:

/home/pietro/nobackup/repo/pandas/pandas/indexes/base.py in _join_non_unique(self, other, how, return_indexers)
   2616         join_index = self.values.take(left_idx)
   2617         mask = left_idx == -1
-> 2618         np.putmask(join_index, mask, other._values.take(right_idx))
   2619 
   2620         join_index = self._wrap_joined_index(join_index, other)

TypeError: putmask() argument 1 must be numpy.ndarray, not Categorical
```
#### Expected Output

No error
#### output of `pd.show_versions()`

```
In [6]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.5.0-2-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.utf8

pandas: 0.18.1+102.g0c6226c
nose: 1.3.7
pip: 1.5.6
setuptools: 18.4
Cython: 0.23.4
numpy: 1.10.4
scipy: 0.16.0
statsmodels: 0.8.0.dev0+0fac349
xarray: None
IPython: 5.0.0.dev
sphinx: 1.3.1
patsy: 0.3.0-dev
dateutil: 2.2
pytz: 2012c
blosc: None
bottleneck: 1.1.0dev
tables: 3.2.2
numexpr: 2.5
matplotlib: 1.5.1
openpyxl: None
xlrd: 0.9.4
xlwt: None
xlsxwriter: 0.7.3
lxml: None
bs4: 4.4.0
html5lib: 0.999
httplib2: 0.9.1
apiclient: 1.5.0
sqlalchemy: 1.0.11
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.38.0
pandas_datareader: 0.2.1
```

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #13365
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

@jreback take this just as a wild guess of how to fix #13365 . If the approach makes sense, I will add tests, whatsnew and some more docs.

### CODE REVIEW

**Comment by jreback** [comment]
you are adding a crazy amount of code here.

File: pandas/indexes/category.py [hunk 1]
```diff
@@ -574,6 +574,35 @@ def append(self, other):
...
         return self._create_from_codes(codes, name=name)
 
+    def _array_putmask(self, a, mask, values):
+        """
+        Equivalent to np.putmask(a, mask, values).
+
+        "a.categories" must already contain all categories that "values" uses.
+        """
+
+        # No vectorized way to build the mapping:
```
Comment (jreback) [inline pandas/indexes/category.py]: ``.get_indexer``

**Comment by toobaz** [comment]
Thanks for the ``.get_indexer()`` suggestion, fixed. Apart from that, do you have in mind a simpler approach? What I _could_ do would be to override the entire ``_join_non_unique()``, duplicating some lines of code but removing the need for the two new methods currently used in the ``CategoricalIndex``- specific parts.

**Comment by toobaz** [comment]
@jreback : what do you want to do?

1) I complete the current approach with tests and docs
2) I override the entire ``_join_non_unique()`` instead that adding two new methods
3) you suggest another approach
4) we close this

**Comment by jreback** [comment]
The approach is not bad, but seems not very elegant. Further your impl of ``_array_putmask`` is not good.

**Comment by toobaz** [comment]
Any _constructive_ criticism? :-) My mood doesn't depend on them, but maybe this PR/issue does.

**Comment by jreback** [comment]
maybe @sinhrks has some

**Comment by sinhrks** [comment]
I may misunderstand, but the point is how pandas handle extended dtype? We should decide whether to make them more numpy compat, or use separated logic using if-else branch. 

I think most of other impl takes latter latter (if-else) option currently.

**Comment by toobaz** [comment]
@sinhrks : yes, the bug is due to a numpy method (``putmask()``) not working (and it's not a mere problem of type checks, there's really missing logic, which this PR introduces) on ``Categorical``.

Certainly the cleanest option (for code organization, not that it would spare us from including the missing logic) would be that ``Categorical`` inherits from np.ndarray. But that's obviously a huge work... and might  (I don't know enough to judge) even be plain impossible due to the in-memory structure of ``Categorical`` being different from a typical ndarray.

Then sure, we can solve this also with "if-else"s without overriding... I have hard time considering such approach more "elegant" than mine, but if it is the path of least resistance to closing the bug I'll update the PR.

**Comment by jreback** [comment]
@toobaz I wasn't looking for a general 'why are we doing things this way', that is obvious. The issue is you are jumping thru hoops and not using the *existing* ``Categorical`` machinery

File: pandas/indexes/base.py [hunk 2]
```diff
@@ -2630,6 +2630,20 @@ def _join_multi(self, other, how, return_indexers=True):
...
+        Must be overridden by Indexes which have non-ndarray .values (e.g.
+        CategoricalIndex).
+        """
+        np.putmask(a, mask, values)
+
+    def _join_values_take_kwargs(self, other):
+        """
+        Must be overridden by Indexes which have non standard
+        self.values.take() keyword args (e.g. CategoricalIndex).
+        """
```
Comment (jreback) [inline pandas/indexes/base.py]: this is really really odd way to do this

File: pandas/indexes/base.py [hunk 3]
```diff
@@ -2630,6 +2630,20 @@ def _join_multi(self, other, how, return_indexers=True):
         raise NotImplementedError("merging with both multi-indexes is not "
                                   "implemented")
 
+    def _array_putmask(self, a, mask, values):
+        """
+        Must be overridden by Indexes which have non-ndarray .values (e.g.
+        CategoricalIndex).
```
Comment (jreback) [inline pandas/indexes/base.py]: this is fine, but the point is to follow existing stype and just call this ``putmask``

File: pandas/indexes/category.py [hunk 4]
```diff
@@ -574,6 +574,32 @@ def append(self, other):
...
+        """
+        Equivalent to np.putmask(a, mask, values).
+
+        "a.categories" must already contain all categories that "values" uses.
+        """
+
+        rep_vals = a.categories.get_indexer(values.categories
+                                            ).view(dtype=a.codes.dtype)
+
+        idces = np.digitize(values.codes, np.arange(len(values.categories)),
```
Comment (jreback) [inline pandas/indexes/category.py]: we have never used ``np.digitize`` in the entire code base. ``CI`` is based on a ``Categorical``. When I said in-elegant I am mainly referring to this impl.

File: pandas/indexes/category.py [hunk 5]
```diff
@@ -574,6 +574,32 @@ def append(self, other):
...
+        new_codes = rep_vals[idces]
+        np.putmask(a.codes, mask, new_codes)
+
+    def _join_values_take_kwargs(self, other):
+        """
+        In a join with CategoricalIndexes, the result of the "take" operation
+        on "self" must be already aware of the new categories it will find in
+        "other".
+        """
+        new_cats = other.categories.difference(self.categories)
```
Comment (jreback) [inline pandas/indexes/category.py]: same here this is a very in-elegant way to do this.

File: pandas/tests/indexes/common.py [hunk 6]
```diff
@@ -225,9 +225,8 @@ def test_copy_name(self):
 
             s1 = Series(2, index=first)
             s2 = Series(3, index=second[:-1])
-            if not isinstance(index, CategoricalIndex):  # See GH13365
-                s3 = s1 * s2
```
Comment (jreback) [inline pandas/tests/indexes/common.py]: you need a LOT more tests

Comment (toobaz) [inline pandas/tests/indexes/common.py]: Yes, _I_ stated this when opening the PR. I will add them, no problem.

File: pandas/indexes/category.py (hunk 4 above)
Comment (toobaz) [inline pandas/indexes/category.py]: To my understanding (which is probably very limited), never in the entire code base do we do anything combining ``Categorical``s with different categories. Am I wrong? Or are you just suggesting I should move this in ``Categorical`` (then yes, I agree with you... but I would still use ``np.digitize``).

File: pandas/indexes/base.py (hunk 2 above)
Comment (toobaz) [inline pandas/indexes/base.py]: ... as already said, if you want me I override the entire ``_join_non_unique`` I will, I don't see many other alternatives currently.

File: pandas/indexes/category.py (hunk 4 above)
Comment (jreback) [inline pandas/indexes/category.py]: see ``union_categoricals``.  I haven't really looked at this, but I don't see any reason to have new categories at all. That violates the guarantee. you would use ``pd.factorize`` instead anyhow. But you should put code in approrpriate places.

I don't have much time to look into this.

File: pandas/indexes/base.py (hunk 3 above)
Comment (toobaz) [inline pandas/indexes/base.py]: Sorry, not following you... but if you're just saying I should change the method name, OK, I will.

**Comment by toobaz** [comment]
OK, now I'm (extending and) using ``union_categorical()`` (thanks for the tip - although I don't have clear what you mean by "violates the guarantee"), and overriding directly ``_join_non_unique()`` so that the change is much less invasive.

**Comment by toobaz** [comment]
@jreback ping

Feel free to close, I won't be offended.

**Comment by jreback** [comment]
yep, don't really have time to look, but needs some refactoring. if you wan to submit again pls feel free.

**Comment by toobaz** [comment]
Submit again?! Well... what?!

I guess the bug(s) will be fixed by somebody who understands how contributing to this project works. I tried in vain.

**Pull request closed without merging**